### PR TITLE
Add calico rock definitions

### PR DESCRIPTION
**Pull request by petrutlucian94**
We're defining rock drop-in replacements for the Calico images.

### CODE REVIEW

**Comment by petrutlucian94** [comment]
@bschimke95 please take a look when you have time.

File: calico/v3.28.0/apiserver/rockcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+description: >
+  This rock is a drop in replacement for the
+  docker.io/calico/apiserver:v3.28.0 image.
+version: v3.28.0
+license: Apache-2.0
+
+base: ubuntu@24.04
+build-base: ubuntu@24.04
+platforms:
+  amd64:
```
Comment (bschimke95) [inline calico/v3.28.0/apiserver/rockcraft.yaml:12]: Why don't we build this ROCK also for ARM?

File: calico/v3.28.0/node-driver-registrar/rockcraft.yaml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    plugin: nil
+    build-snaps:
+      - go/1.22/stable
+
+  node-driver-registrar:
+    after: [build-deps]
+    plugin: go
+    source-type: git
+    source: https://github.com/kubernetes-csi/node-driver-registrar
+    source-tag: master
```
Comment (bschimke95) [inline calico/v3.28.0/node-driver-registrar/rockcraft.yaml]: should we pin this?

File: calico/v3.28.0/node/rockcraft.yaml [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+      - libtool
+      - llvm
+      - make
+      - libpcre3-dev
+      - pkg-config
+      - protobuf-compiler
+      - wget
+      - zip
+      - libelf-dev
+      # - libbpf-dev
```
Comment (bschimke95) [inline calico/v3.28.0/node/rockcraft.yaml:59]: why is this commented out?

File: calico/v3.28.0/node-driver-registrar/rockcraft.yaml (hunk 2 above)
Comment (petrutlucian94) [reply calico/v3.28.0/node-driver-registrar/rockcraft.yaml]: It's actually pinned:

```
override-build: |
       git checkout a498c31b88cd31f8ea7bafd0d715a75f9939fa1f
```

The git plugin doesn't allow commit hashes to be specified with the ``source`` label.

File: calico/v3.28.0/node/rockcraft.yaml (hunk 3 above)
Comment (petrutlucian94) [reply calico/v3.28.0/node/rockcraft.yaml:59]: We had to build it ourselves due to a version compatibility issue:

```
  libbpf:
     after: [build-deps]
     source: https://github.com/libbpf/libbpf
     source-tag: v0.6.1
     source-type: git
     source-depth: 1
     plugin: nil
     override-build: |
       # Calico v3.28.0 isn't compatible with libbpf>=1.0.
       # v0.7.0 has deprecated some helpers.
```

3.29.0 and onwards may use the prebuilt package by uncommenting that line and dropping the libbpf section.

File: tigera-operator/v1.34.0/rockcraft.yaml [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+      - go/1.22/stable
+
+  tigera-operator:
+    after: [build-deps]
+    plugin: go
+    source-type: git
+    source: https://github.com/tigera/operator
+    source-tag: v1.34.0
+    source-depth: 1
+    override-build: |
```
Comment (berkayoz) [inline tigera-operator/v1.34.0/rockcraft.yaml:39]: Can we keep the patches as `.patch` files and import/use them as we do with entrypoint scripts?

File: calico/v3.28.0/apiserver/rockcraft.yaml (hunk 1 above)
Comment (petrutlucian94) [reply calico/v3.28.0/apiserver/rockcraft.yaml:12]: Based on the previous discussions, I thought ARM wasn't needed for the time being. Also, I'm not sure if I have access to the ARM job runners.

File: calico/v3.28.0/node/rockcraft.yaml [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+      - nftables
+      - ipset
+      - iproute2
+      - libpcap0.8t64
+      - libcryptsetup12
+      - gzip
+      - net-tools
+      - kmod
+      - procps
+      # conntrack
```
Comment (berkayoz) [inline calico/v3.28.0/node/rockcraft.yaml]: Is this not needed at runtime?

File: calico/v3.28.0/apiserver/rockcraft.yaml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+  apiserver:
+    after: [build-deps]
+    plugin: go
+    source-type: git
+    source: https://github.com/projectcalico/calico
+    source-tag: v3.28.0
+    source-depth: 1
+    override-build: |
+      PACKAGE_NAME=github.com/projectcalico/calico/apiserver
```
Comment (berkayoz) [inline calico/v3.28.0/apiserver/rockcraft.yaml]: Can we move constant build environment variables to the `build-environment` part property?
Such as `PACKAGE_NAME`, `GIT_VERSION`, `CGO_ENABLED`

File: tests/integration/test_calico.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    return {
+        "registry": parts[0],
+        "repo": parts[1],
+        "name": parts[2].split(":")[0],
+        "tag": parts[2].split(":")[1],
+    }
+
+
+@pytest.mark.parametrize("operator_version,calico_version", [("v1.34.0", "v3.28.0")])
+def test_calico(
```
Comment (berkayoz) [inline tests/integration/test_calico.py:133]: I believe we are not disabling the network feature of the k8s-snap which is cilium by default, I feel running 2 CNIs could lead to false positives. Should we disable the feature with a bootstrap config?

We will be testing these images more extensively when moonray branch switches the images.

**Review by berkayoz — COMMENTED** [review]
Great work, left some comments

File: calico/v3.28.0/node-driver-registrar/rockcraft.yaml (hunk 2 above)
Comment (berkayoz) [reply calico/v3.28.0/node-driver-registrar/rockcraft.yaml]: `source-commit` should work unless I'm missing something, check [the docs ](https://documentation.ubuntu.com/rockcraft/en/latest/common/craft-parts/reference/part_properties/#source-commit).

Comment (petrutlucian94) [reply calico/v3.28.0/node-driver-registrar/rockcraft.yaml]: > source-commit should work unless I'm missing something, check the docs .

Missed that, thanks!

File: tigera-operator/v1.34.0/rockcraft.yaml (hunk 4 above)
Comment (petrutlucian94) [reply tigera-operator/v1.34.0/rockcraft.yaml:39]: Good point!

File: calico/v3.28.0/apiserver/rockcraft.yaml (hunk 1 above)
Comment (berkayoz) [reply calico/v3.28.0/apiserver/rockcraft.yaml:12]: I believe we are building the `-moonray` flavor for arm64. I think building this for ARM might be better. For ARM runners we'd need to use/move to a branch from this repo

File: calico/v3.28.0/apiserver/rockcraft.yaml (hunk 6 above)
Comment (petrutlucian94) [reply calico/v3.28.0/apiserver/rockcraft.yaml]: Sure. Having ``CGO_ENABLED=0`` right next to the build command may improve visibility but it's probably ok either way.

File: calico/v3.28.0/node/rockcraft.yaml (hunk 5 above)
Comment (petrutlucian94) [reply calico/v3.28.0/node/rockcraft.yaml]: It's actually included a few lines below. The reference image had an inline comment that separated conntrack dependencies, which seemed like a good idea.

I can change the comment to ``# conntrack dependencies``.

File: tests/integration/test_calico.py (hunk 7 above)
Comment (petrutlucian94) [reply tests/integration/test_calico.py:133]: Right, I didn't want to mingle with the k8s snap bootstrap process, so I set Calico to be a no-op:

```
apiVersion: operator.tigera.io/v1
 kind: Installation
   ...
   calicoNetwork:
     ipPools:
     - name: default-ipv4-ippool
       blockSize: 26
       cidr: 192.168.0.0/16
       encapsulation: VXLANCrossSubnet
       natOutgoing: Disabled
       nodeSelector: !all()
```

File: calico/v3.28.0/node/rockcraft.yaml (hunk 3 above)
Comment (petrutlucian94) [reply calico/v3.28.0/node/rockcraft.yaml:59]: I'll add an inline comment.

File: calico/v3.28.0/apiserver/rockcraft.yaml (hunk 1 above)
Comment (petrutlucian94) [reply calico/v3.28.0/apiserver/rockcraft.yaml:12]: > we'd need to use/move to a branch from this repo

I can't do that either. However, I got access to a Cloudbase ARM machine so I built the images there.

Comment (bschimke95) [reply calico/v3.28.0/apiserver/rockcraft.yaml:12]: Why can't you do that? You should be part of the Canonical Kubernetes team and hence have the permissions.

Comment (petrutlucian94) [reply calico/v3.28.0/apiserver/rockcraft.yaml:12]: My request to join the Github Canonical organization hasn't been approved yet.
image

**Review by berkayoz — APPROVED** [review]

**Pull request merged**